src/optionals/log.c

```diff
@@ -0,0 +1,298 @@
+        return EMPTY_VAL;
+    }
+
+    Log *log = AS_LOG(args[0]);
+    log->prefix = strdup("");
+
+    return NIL_VAL;
+}
+
+static Value logObjPrint(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "print() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Log *log = AS_LOG(args[0]);
+
+    struct tm *ptm = timeNow();
+    char *msg = AS_CSTRING(args[1]);
+
+    if (log->prefix != NULL && strlen(log->prefix) != 0) {
+        fprintf(log->of, "%04d/%02d/%02d %02d:%02d:%02d %s %s", 
+            ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+            ptm->tm_hour, ptm->tm_min, ptm->tm_sec, log->prefix, msg);
+
+        return NIL_VAL;
+    }
+
+    fprintf(log->of, "%04d/%02d/%02d %02d:%02d:%02d %s", 
+        ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+        ptm->tm_hour, ptm->tm_min, ptm->tm_sec, msg);
+
+    return NIL_VAL;
+}
+
+static Value logObjPrintln(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "println() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Log *log = AS_LOG(args[0]);
+
+    struct tm *ptm = timeNow();
+    char *msg = AS_CSTRING(args[1]);
+
+    if (log->prefix != NULL && strlen(log->prefix) != 0) {
+        fprintf(log->of, "%04d/%02d/%02d %02d:%02d:%02d %s %s\n", 
+            ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+            ptm->tm_hour, ptm->tm_min, ptm->tm_sec, log->prefix, msg);
+
+        return NIL_VAL;
+    }
+
+    fprintf(log->of, "%04d/%02d/%02d %02d:%02d:%02d %s\n", 
+        ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+        ptm->tm_hour, ptm->tm_min, ptm->tm_sec, msg);
+
+    return NIL_VAL;
+}
+
+static Value logObjFatal(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "fatal() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Log *log = AS_LOG(args[0]);
+
+    struct tm *ptm = timeNow();
+
+    char *msg = AS_CSTRING(args[1]);
+    fprintf(log->of, "%04d/%02d/%02d %02d:%02d:%02d %s\n", 
+        ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+        ptm->tm_hour, ptm->tm_min, ptm->tm_sec, msg);
+
+    exit(1);
+}
+
+static Value logObjFatalln(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "fatalln() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Log *log = AS_LOG(args[0]);
+
+    struct tm *ptm = timeNow();
+
+    char *msg = AS_CSTRING(args[1]);
+    fprintf(log->of, "%04d/%02d/%02d %02d:%02d:%02d %s\n", 
+        ptm->tm_year+2000-100, ptm->tm_mon, ptm->tm_mday, 
+        ptm->tm_hour, ptm->tm_min, ptm->tm_sec, msg);
+
+    exit(1);
+}
+
+ObjAbstract* newLogObj(DictuVM *vm) {
+    ObjAbstract *abstract = newAbstract(vm, freeLog);
+    push(vm, OBJ_VAL(abstract));
+
+    Log *log = ALLOCATE(vm, Log, 1);
```

Comment:
```suggestion
    Log *log = ALLOCATE(vm, Log, 1);
    log->of = NULL;
    log->prefix = NULL;
```

We will want to initialise both of those to NULL. If a prefix is never set for example, when trying to free a Log object we're checking for NULL, we want to ensure that it's actually NULL rather than a garbage value.